firedrake/slate/slac/kernel_builder.py

```diff
@@ -1,19 +1,39 @@
-
-from collections import OrderedDict
-
 from coffee import base as ast
 
-from firedrake.slate.slate import (TensorBase, Tensor, TensorOp,
-                                   Action, Inverse)
-from firedrake.slate.slac.utils import (Transformer, traverse_dags,
-                                        collect_reference_count)
+from collections import OrderedDict, Counter, namedtuple
+
+from firedrake.slate.slac.utils import (topological_sort, traverse_dags,
+                                        eigen_tensor, Transformer)
 from firedrake.utils import cached_property
 
+from functools import reduce
+
 from ufl import MixedElement
 
+import firedrake.slate.slate as slate
+
+
+CoefficientInfo = namedtuple("CoefficientInfo",
+                             ["space_index",
+                              "offset_index",
+                              "shape",
+                              "coefficient"])
+CoefficientInfo.__doc__ = """\
+Context information for creating coefficient temporaries.
 
-class KernelBuilder(object):
-    """A helper class for constructing Slate kernels.
+:param space_index: An integer denoting the function space index.
+:param offset_index: An integer denoting the starting position in
+                     the vector temporary for assignment.
+:param shape: A singleton with an integer describing the shape of
+              the coefficient temporary.
+:param coefficient: The :class:`ufl.Coefficient` containing the
+                    relevant data to be placed into the temporary.
+"""
+
+
+class LocalKernelBuilder(object):
+    """The primary helper class for constructing cell-local linear
```

Comment:
I decided to not split the builder class after all, that's the only real change the last couple of commits did.